### ISSUE TITLE
Hyprland WM is not smooth on intel-iGPU

### ISSUE THREAD

**Issue by joelmathewthomas** [hyprwm/Hyprland#5441]
### Hyprland Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch HEAD at commit 3875679755014997776e091ff8903acfb311dd2f  (props: bump ver to 0.38.0).
Date: Mon Apr 1 18:30:37 2024
Tag: v0.38.0

flags: (if any)


System Information:
System name: Linux
Node name: nitro5
Release: 6.8.2-arch2-1
Version: #1 SMP PREEMPT_DYNAMIC Thu, 28 Mar 2024 17:06:35 +0000


GPU information: 
0000:00:02.0 VGA compatible controller [0300]: Intel Corporation TigerLake-H GT1 [UHD Graphics] [8086:9a60] (rev 01) (prog-if 00 [VGA controller])
0000:01:00.0 VGA compatible controller [0300]: NVIDIA Corporation GA107M [GeForce RTX 3050 Ti Mobile] [10de:25a0] (rev a1) (prog-if 00 [VGA controller])


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:

```

</details>

### Bug or Regression?

Bug

### Description

So, i just installed hyprland on a multi-gpu laptop. The first thing i  noticed is that , the it has very occasional and inconsistent framedrops. 

 My display is 144hz. And this is my monitor Config. `monitor=,highrr,auto,1`.

 So to give you a better understanding. This is what is the case when i first login to hyprland session.
![photo1712292421(1)](https://github.com/hyprwm/Hyprland/assets/90510078/f22ac864-9791-4c48-adba-d0da8994805d)

But after opening a few windows or moving them around, basically just working the hyprland animations. This is what i get.

![photo1712292421](https://github.com/hyprwm/Hyprland/assets/90510078/c05767c4-55fe-4761-bb5e-fda36dfdd5d2)

So, basically the issue is that , there is always inconsistent framedrops. It just is not smooth. 
If anyone knows a fix or solution, please reply.

 
To make sure everything is rendered on the iGPU, i ran this

```glxinfo | grep "OpenGL renderer"
OpenGL renderer string: Mesa Intel(R) UHD Graphics (TGL GT1)
```

and

```sudo inxi -G
Graphics:
  Device-1: Intel TigerLake-H GT1 [UHD Graphics] driver: i915 v: kernel
  Device-2: NVIDIA GA107M [GeForce RTX 3050 Ti Mobile] driver: nvidia
    v: 550.67
  Device-3: Chicony HD User Facing driver: uvcvideo type: USB
  Display: server: X.org v: 1.21.1.12 with: Xwayland v: 23.2.5 driver: X:
    loaded: modesetting,nvidia dri: iris gpu: i915 resolution: 1920x1080
  API: EGL v: 1.5 drivers: iris,kms_swrast,nvidia,swrast
    platforms: gbm,surfaceless,device
  API: OpenGL v: 4.6.0 compat-v: 4.5 vendor: mesa v: 24.0.4-arch1.2
    note: incomplete (EGL sourced) renderer: llvmpipe (LLVM 17.0.6 256 bits),
    NVIDIA GeForce RTX 3050 Ti Laptop GPU/PCIe/SSE2, Mesa Intel UHD Graphics
    (TGL GT1)
```

### How to reproduce

Install Hyprland on a Multi GPU system with integrated intel GPU, and dedicated NVIDIA GPU.

### Crash reports, logs, images, videos

No crash reports

**Comment by vaxerski** [comment]
could be #5426?

Have you checked if this happens on an older version, e.g. 0.36?

**Comment by joelmathewthomas** [comment]
Yes , I have checked. I'm not sure about the version. But I had the same issue 4 months ago. So I decided to stay off until it hopefully got fixed. But unfortunately, it's still there.

**Comment by vaxerski** [comment]
generally due to how wlroots' frame scheduler works, if your gpu is unable to render a frame within the frametime window, the framerate will halve.

Though I don't see why opening a few windows in the background would be putting any significant additional strain on the gpu.

**Comment by joelmathewthomas** [comment]
Exactly. I don't understand as well. When I start the session everything works fine. But then if I open something like thunar, and just by moving windows i can see it's not smooth, feels more like 60 or 70 fps, when it should be 144fps. 

If there's anything you like to know, like a log, or anything you'd like me to check out, please do tell me.

**Comment by vaxerski** [comment]
if you are moving windows at the same time as the test is running, it's normal it will tank - rendering the movement of the window is a lot of pixels.

**Comment by joelmathewthomas** [comment]
> if you are moving windows at the same time as the test is running, it's normal it will tank - rendering the movement of the window is a lot of pixels.

Okay, that's expected. But I think the same happened even when the test wasn't running. But please allow me  to confirm it once I get back to my system.

Is there any possible way I could see an fps monitor, so I can actually determine the extent of performance loss.

**Comment by vaxerski** [comment]
set `debug:overlay` to true https://wiki.hyprland.org/Configuring/Variables/#debug

**Comment by vaxerski** [comment]
yeah odd. check 5426 as I feel like this could be the same issue.

Try reverting to 0.36-ish and check if it happens there too

**Comment by joelmathewthomas** [comment]
Yeah , I'll do that

**Comment by joelmathewthomas** [comment]
I tried out version 0.36 , installed it using cmake.

```hyprctl systeminfo
Hyprland, built from branch HEAD at commit 1c460e98f870676b15871fe4e5bfeb1a32a3d6d8  (props: bump ver to 0.36.0).
Date: Wed Feb 28 00:32:40 2024
Tag: v0.36.0

flags: (if any)


System Information:
System name: Linux
Node name: nitro5
Release: 6.8.2-arch2-1
Version: #1 SMP PREEMPT_DYNAMIC Thu, 28 Mar 2024 17:06:35 +0000


GPU information: 
0000:00:02.0 VGA compatible controller [0300]: Intel Corporation TigerLake-H GT1 [UHD Graphics] [8086:9a60] (rev 01) (prog-if 00 [VGA controller])
0000:01:00.0 VGA compatible controller [0300]: NVIDIA Corporation GA107M [GeForce RTX 3050 Ti Mobile] [10de:25a0] (rev a1) (prog-if 00 [VGA controller])


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:
```

Unfortunately, the frame drops are still there. Even a small animation like just moving a window, drops the fps from 144fps to weirdly exactly 72 fps. 
But one thing that's different from before , is once i don't do anything, i mean with animations, the fps returns to 144 fps, which it did not with version 0,38(where it just stayed at low fps)..

**Comment by joelmathewthomas** [comment]
> read my previous comments. That's expected. 

I wasn't talking about running glxgears.
I was saying , without any test running, and just only hyprland, I monitored the fps using `debug: overlay` and it's dropping from 144fps to 72fps. 

I just don't feel like it's expected since, I have a low end laptop with amd igpu and a 60hz display, and the fps is pretty stable.

> please help us over there by bisecting this if you have a bit of spare time.

Will definitely do it tomorrow. 
Thank you for your time and efforts.

**Comment by GokuPlay609** [comment]
I am experiencing same issue sadly there is no info about this

**Comment by joelmathewthomas** [comment]
> I am experiencing same issue sadly there is no info about this

You could try using intel_gpu_frequency to set your GPU frequency to max.

[intelgpufreq.service.docx](https://github.com/user-attachments/files/20060132/intelgpufreq.service.docx)

I've attached a document, mentioning how to create a systemd service to do it on boot. Note that the text formatting in the document might be weird.

More info over at #5456

**Comment by boudywho** [comment]
> > I am experiencing same issue sadly there is no info about this
> 
> You could try using intel_gpu_frequency to set your GPU frequency to max.
> 
> [intelgpufreq.service.docx](https://github.com/user-attachments/files/20060132/intelgpufreq.service.docx)
> 
> I've attached a document, mentioning how to create a systemd service to do it on boot. Note that the text formatting in the document might be weird.
> 
> More info over at [#5456](https://github.com/hyprwm/Hyprland/issues/5456)

it kind of helps but it doesn't fully remove stutters and lags, do you have any other idea on how to improve the performance?
I have intel igpu

**Comment by joelmathewthomas** [comment]
> > > I am experiencing same issue sadly there is no info about this
> > 
> > 
> > You could try using intel_gpu_frequency to set your GPU frequency to max.
> > [intelgpufreq.service.docx](https://github.com/user-attachments/files/20060132/intelgpufreq.service.docx)
> > I've attached a document, mentioning how to create a systemd service to do it on boot. Note that the text formatting in the document might be weird.
> > More info over at [#5456](https://github.com/hyprwm/Hyprland/issues/5456)
> 
> it kind of helps but it doesn't fully remove stutters and lags, do you have any other idea on how to improve the performance? I have intel igpu

https://wiki.hyprland.org/Configuring/Performance/#low-fpsstutterfps-drops-on-intel-igpu-with-tlp-mainly-laptops

You could try this out.

**Comment by Evgeniycity** [comment]
Same problem with msi katana gf76

**Comment by nagualcode** [comment]
try if:
'echo 500 > /sys/class/drm/card1/gt_min_freq_mhz'
works
(this temporary and not persistent across reboots)
